### ISSUE TITLE
'numpy.core._multiarray_umath' not found on embedded Python, but found on console.

### ISSUE THREAD

**Issue by RichardEdye** [numpy/numpy#13051]
I can import numpy in the Python console, but cannot import it in an embedded python C++ application that used the same Python version with the same folders. I have only this Python version on my PC. Importing other packages from the application works. 

The application sets Py_SetPythonHome() to my Python folder: 
C:\\Users\\MyName\\AppData\\Local\\Programs\\Python\\Python37-32

print(os.__file__) prints the correct file.

### Reproducing code example:

```C++
Py_SetProgramName(Py_DecodeLocale(ProgramName,0));
Py_SetPythonHome(Py_DecodeLocale("C:\\Users\\Richard\\AppData\\Local\\Programs\\Python\\Python37-32",0));
Py_Initialize();
PyModuleMain = PyImport_AddModule("__main__");
PyGlobalDict = PyModule_GetDict(PyModuleMain);
const char *Content = "import os; print(os.__file__)\nimport numpy";
PyRun_SimpleString(Content);
```

```python
import os; print(os.__file__)
import numpy
```
### Error message:

Traceback (most recent call last):
  File "C:\Users\Richard\AppData\Local\Programs\Python\Python37-32\lib\site-packages\numpy\core\__init__.py", line 40, in <module>
    from . import multiarray
  File "C:\Users\Richard\AppData\Local\Programs\Python\Python37-32\lib\site-packages\numpy\core\multiarray.py", line 12, in <module>
    from . import overrides
  File "C:\Users\Richard\AppData\Local\Programs\Python\Python37-32\lib\site-packages\numpy\core\overrides.py", line 6, in <module>
    from numpy.core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 2, in <module>
  File "C:\Users\Richard\AppData\Local\Programs\Python\Python37-32\lib\site-packages\numpy\__init__.py", line 142, in <module>
    from . import core
  File "C:\Users\Richard\AppData\Local\Programs\Python\Python37-32\lib\site-packages\numpy\core\__init__.py", line 71, in <module>
    raise ImportError(msg)
ImportError: 
...

### Numpy/Python version information:

1.16.2, 3.7 32 bit Windows

**Comment by charris** [comment]
This is probably a duplicate of several other issues but I think the embedded Python is new. See if the discussion at #12977. How did you install NumPy? Are you working with Anaconda?

**Comment by RichardEdye** [comment]
I have installed with pip install numpy. No Anaconda and no other Pythons on the system. 

I get the error message with the Python debug DLL. When I use the release DLL, the line "import numpy" immediately causes a crash in the PyRun_SimpleString() call.

**Comment by RichardEdye** [comment]
I have also added  site-packages\\numpy\\.libs to the PATH as suggested in the other thread, but to no avail. That folder contains a libopenblas.JKAMQ5EVHIVCPXP2XZJB2RQPIN47S32M.gfortran-win32.dll.

**Comment by mattip** [comment]
A couple of things might be happening. 
- Something is affecting the [dll search order](https://docs.microsoft.com/en-us/windows/desktop/Dlls/dynamic-link-library-search-order). This might be offset by copying the dll from`'.lib` to `core` next to the `_multiarray_umath.pyd` file.
- Something in your embedded app uses a different runtime library which prevents loading numpy. 

> Importing other packages from the application works

Can you use other c-extension modules, for instance lxml?

**Comment by RichardEdye** [comment]
Yes, importing lxml works with my app. And copying the DLL from .libs to core did not help. BTW, the pyd file in core is named "_multiarray_umath.cp36-win32.pyd".

If it helps, I can upload my app for testing.

**Comment by mattip** [comment]
a link to the app repo or a pastebin with the code might be helpful, but I don't really use Windows these days so it might take a while to look at.

**Comment by elikosan** [comment]
I have the same problem. I have an application with embedded python. When trying to load numpy, it fails with the same error. And it works fine from the command line python.

I found that it also fails to load pytorch, with a similar error. The common issue i found is that the failed files have an extension made by concatenating the python version and the platform (e.g. cp37-win_amd64). 

Since the PATH are fine, i think the problem is that the embedded python does not know how to build that extension. I'm trying to dig this out, but to no avail so far. But it seems that  the problem is in Python rather than numpy (or pytorch).


In numpy, the failed command is:
from . import _multiarray_umath
and the corresponding file is:
C:\Program Files\Python37\Lib\site-packages\numpy\core\_multiarray_umath.cp37-win_amd64.pyd

In pytorch, the failed command is:
from torch._C import *
and the corresponding file is:
C:\Program Files\Python37\Lib\site-packages\torch\_C.cp37-win_amd64.pyd

**Comment by mattip** [comment]
Is there a missing `\` between `core` and `_multiarray_umath`? The `pyd` file should sit in the `core` directory.

**Comment by elikosan** [comment]
Yes. The github web editor seems to eat the backslash when rendering a backslash followed by an underscore. I added a space.

**Comment by mattip** [comment]
> Install Python 3.6. and Numpy on that PC ...

Are you mixing python 3.6 and python 3.7?

**Comment by RichardEdye** [comment]
I used 3.7 first, but switched back to 3.6 for checking if this would solve the problem - which it didn't.

**Comment by elikosan** [comment]
My problem was that I was building the embedded python in Debug mode. In this case python appends a _d to the suffixes when loading files. However, numpy installation did not include those files, so the import would fail.

**Comment by RichardEdye** [comment]
Yes, but that's not the real problem. When building it in release mode, the PyRun_SimpleString(...) call will crash when numpy is imported. This crash seems to depend on circumstances, for instance loading other libraries. Sometimes it does not crash on the first, but always on the second call.

**Comment by cgohlke** [comment]
FWIW, I cannot reproduce this with numpy-1.16.2 or 1.16.4, Python 3.7.3, Visual Studio 2017.

**Comment by RichardEdye** [comment]
With Python 3,7 the effect is a bit different than with 3.6. It imports successfully on the first call, and crashes on subsequent calls. You can test this by adding to my code example above some exit code

Py_Finalize();
PyMem_RawFree(Exe);

and then calling it twice from the main program.

**Comment by andimarafioti** [comment]
Hey, I had this same problem on windows 10 with Python 3.6. Solved it by upgrading to numpy 1.16.4 (had 1.13 something).

**Comment by dlong11** [comment]
I am seeing this null pointer reference on Windows 10 with version 1.16.3 and 1.16.4. (python 3.6 64-bit)

Simple example 
```
Py_SetProgramName(Py_DecodeLocale("Test",0));
Py_Initialize();
PyImport_AddModule("__main__");
const char* content = "import os; print(os.__file__)\nimport numpy";
PyRun_SimpleString(content);
Py_Finalize();

Py_Initialize();
PyImport_AddModule("__main__");
PyRun_SimpleString(content);
Py_Finalize();
```

Exception thrown at 0x00007FFAE7C1DC19 (_multiarray_umath.cp36-win_amd64.pyd) in test.exe: 0xC0000005: Access violation writing location 0x000000000000000A. occurred

**Comment by eric-wieser** [comment]
@dlong11: your problem is a different one, reimporting numpy multiple times in different embedded interpreters is not supported. You'll likely find other issues about that.

**Comment by RichardEdye** [comment]
But that's indeed the problem that I also have. The newer Numpy versions don't give the error message anymore, but they run only once and crash on the second run. I must then close the application and open it again. Seems Numpy does not do a clean exit or does not release some allocated resource.

Where can I find information about it? Shall I open a second issue?

**Comment by dlong11** [comment]
@eric-wieser Thanks for the info.  That stinks.  ☹️  Is this something that "NumPy" cares about?  Looking for volunteers?   Being able tp restart the interpreter is pretty important when embedding python.

**Comment by eric-wieser** [comment]
Historically, numpy doesn't care enough about it to spend any time worrying about it, but patches welcome.

Personally, I'm not at all sure what fixing it would entail - at a guess, we'd need to ensure we drop every PyObject reference at shutdown, which includes ones stored in function-static variables?

This might be easier to fix now that master tracks python 3.5 and onward than it was when I last looked at it.

**Comment by SHITINGCHEN** [comment]
Have u checked if it's the release or debug problem? I've met the same situation as you did, using python 3.7.4 and numpy 1.17 (installed through pip) on windows. In C++ program, check sys.path of embedded python to see if it is looking for a "_d" suffix. And changing from debug mode to release mode solved my problem.

**Comment by RichardEdye** [comment]
It's a release problem. I haven't tested it with 1.17, yet, but will do in the next days.

**Comment by mattip** [comment]
Closing. This issue started about trying to use a release version of NumPy on a debug version of windows python, and transitioned to talking about reimporting after calling `Py_Finalize();`, which is not supported.

Please open a new issue if there is something I missed.

**Comment by RichardEdye** [comment]
Well, I fear it has absoltuely nothing to do with debug or release versions. But this is a quite old thread and my software was delivered long ago. So for me the issue is closed. 

I just had to write in the documentation that NumPy does not clean exit, so for starting it twice the user had to close and restart the whole application. Fortunately, only Numpy had this problem, the other python libs that I tested worked so far.

**Comment by alkalinin** [comment]
This is a problem of the 'pyd' naming convention, see [python issue 36716](https://bugs.python.org/issue36716).

```python
import importlib.machinery
print(importlib.machinery.all_suffixes())
```  
CommandLine Python:
```python
> ['.py', '.pyw', '.pyc', '.cp37-win_amd64.pyd', '.pyd']
```
C++ Embedded Python:
```python
> ['.py', '.pyw', '.pyc', '_d.cp37-win_amd64.pyd', '_d.pyd']
```

**Comment by spartha77** [comment]
I too had the same problem. I solved it.
My problem statement: (Windows)

-> I had installed Python3.5 at C:\Python35, 
-> I invoked Python and ran "import numpy" 
It gave rise to similar problem.
<< I then realized that the numpy installation was at C:\Users\partha3x\AppData\Roaming\Python\Python37\site-packages\numpy\core\ and NOT from the path I had mentioned at the time of installation. 
-> So, I modified (re-installed) python at the user profile home directory.
-> The home profile directory is 
C:\Users\partha3x\AppData\Local\Programs\Python\Python37
->Uninstalled the numpy 
->Re-installed numpy
With this "import numpy" call succeeded in python 

With Visual Studio 2019, I had the same problem. I followed the following steps
-> Uninstalled all Pythons that which were installed using VIsual Studio installer. (Not sure if this step helped me)
-> In the project properties, in which I was invoking Python (embedded Python) from C++ project, I had set additional Libraries ( Project properties->Linker->General->Additional Library Directories I added C:\Users\partha3x\AppData\Roaming\Python\Python37\site-packages\numpy\core\lib AND -> In the project properties -> Linker->Input->Additional Dependencies , I added npymath.lib.

These fixed the issues. 

In Linux, similar to the project properties, the files have to be compiled with additional dependencies with additional directories using make files.

Hope this helps.

**Comment by MyFirstGitEver** [comment]
> My problem was that I was building the embedded python in Debug mode. In this case python appends a _d to the suffixes when loading files. However, numpy installation did not include those files, so the import would fail.

I'm also stuck with this problem. Could you help me? What do you mean by "I was building the embedded python in Debug mode". How can I change that? I tried to switch to release mode and the include lines just turn red for some reason. They wont let me compile

**Comment by MyFirstGitEver** [comment]
> Have u checked if it's the release or debug problem? I've met the same situation as you did, using python 3.7.4 and numpy 1.17 (installed through pip) on windows. In C++ program, check sys.path of embedded python to see if it is looking for a "_d" suffix. And changing from debug mode to release mode solved my problem.

I switched to release mode. Now Py_initialize throws "Could not find platform independent libraries <prefix"
